backends/utils/history_server.go

```
@@ -0,0 +1,363 @@
+				_ = AppendColumn(defaultColumns[4], entry.StartTime)
+				_ = AppendColumn(defaultColumns[5], entry.ActionType)
+				_ = AppendColumn(defaultColumns[6], entry.Container)
+
+				for k, v := range entry.AdditionalData {
+					// If this column already exists, append new data
+					if _, ok := customColumnsByName[k]; !ok {
+						// Create a new column with nil values up until this row
+						data := make([]string, i)
+						data[i-1] = v
+						customColumnsByName[k], err = frames.NewSliceColumn(k, data)
+						if err != nil {
+							return nil, err
+						}
+					} else {
+						_ = AppendColumn(customColumnsByName[k], v)
+					}
+				}
+
+				// Add null values
+				for columnName, col := range customColumnsByName {
+					if _, ok := entry.AdditionalData[columnName]; !ok {
+						_ = AppendNil(col)
+					}
+				}
+
+				_ = AppendColumn(indexColumn, i)
+				i++
+			}
+		}
+
+		if scanner.Err() != nil {
+			return nil, err
+		}
+
+		// in case previous chunk ended exactly in the end of a row and we still got more data to process
+		if fileIterator.Next() {
+			goto ScanningFileChunk
+		} else if fileIterator.Error() != nil {
+			return nil, fileIterator.Error()
+		}
+	}
+
+	if iter.Err() != nil {
+		return nil, iter.Err()
+	}
+	allColumns := defaultColumns
+
+	for _, col := range customColumnsByName {
+		allColumns = append(allColumns, col)
+	}
+
+	return frames.NewFrame(allColumns, []frames.Column{indexColumn}, nil)
+}
+
+func (m *HistoryServer) writeMonitoringBatch(logs []HistoryEntry) {
+	d, err := json.Marshal(logs)
+	if err != nil {
+		m.logger.Warn("Failed to marshal logs to json, err: %v. logs: %v", err, logs)
+		return
+	}
+
+	d = append(d, []byte("\n")...)
+	input := &v3io.PutObjectInput{Path: m.logFilePath, Body: d, Append: true}
+	err = m.container.PutObjectSync(input)
+	if err != nil {
+		m.logger.Warn("Failed to append Frames History logs to file, err: %v. logs: %v", err, logs)
```

I'm not sure discarding the logs and moving on is the right thing to do. It's not like it's the logs' fault.

---

well i tried to do the entire "error handling" of the ingestion of the logs kinda seamless. because:
1. its async and not part of a user action where you can return an error.
2. it shouldn't effect any normal use of the service. hence i thought for now just logging it should be good enough.

do you have other suggestions?